pkg/controller/volume/vacprotection/vac_protection_controller_test.go

```diff
@@ -0,0 +1,355 @@
+		initialObjects []runtime.Object
+		// Optional client reactors.
+		reactors []reaction
+
+		// VAC event to simulate. This VAC will be automatically added to
+		// initialObjects.
+		updatedVAC *storagev1alpha1.VolumeAttributesClass
+
+		// PV event to simulate. The updatedPV will be automatically added to
+		// initialObjects.
+		oldPV     *v1.PersistentVolume
+		updatedPV *v1.PersistentVolume
+
+		// PVC event to simulate. The updatedPVC will be automatically added to
+		// initialObjects.
+		oldPVC     *v1.PersistentVolumeClaim
+		updatedPVC *v1.PersistentVolumeClaim
+
+		// List of expected kubeclient actions that should happen during the
+		// test.
+		expectedActions []clienttesting.Action
+	}{
+		// VAC events
+		//
+		{
+			name:       "VAC without finalizer -> finalizer is added",
+			updatedVAC: vac1,
+			expectedActions: []clienttesting.Action{
+				clienttesting.NewUpdateAction(vacGVR, "", vac1WithFinalizer),
+			},
+		},
+		{
+			name:            "VAC with finalizer -> no action",
+			updatedVAC:      vac1WithFinalizer,
+			expectedActions: []clienttesting.Action{},
+		},
+		{
+			name:       "saving VAC finalizer fails -> controller retries",
+			updatedVAC: vac1,
+			reactors: []reaction{
+				{
+					verb:      "update",
+					resource:  "volumeattributesclasses",
+					reactorfn: generateUpdateErrorFunc(t, 2 /* update fails twice*/),
+				},
+			},
+			expectedActions: []clienttesting.Action{
+				// This fails
+				clienttesting.NewUpdateAction(vacGVR, "", vac1WithFinalizer),
+				// This fails too
+				clienttesting.NewUpdateAction(vacGVR, "", vac1WithFinalizer),
+				// This succeeds
+				clienttesting.NewUpdateAction(vacGVR, "", vac1WithFinalizer),
+			},
+		},
+		{
+			name:       "deleted VAC with finalizer -> finalizer is removed",
+			updatedVAC: vac1TerminatingWithFinalizer,
+			expectedActions: []clienttesting.Action{
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+			},
+		},
+		{
+			name:       "finalizer removal fails -> controller retries",
+			updatedVAC: vac1TerminatingWithFinalizer,
+			reactors: []reaction{
+				{
+					verb:      "update",
+					resource:  "volumeattributesclasses",
+					reactorfn: generateUpdateErrorFunc(t, 2 /* update fails twice*/),
+				},
+			},
+			expectedActions: []clienttesting.Action{
+				// Fails
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+				// Fails too
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+				// Succeeds
+				clienttesting.NewUpdateAction(vacGVR, "", vac1Terminating),
+			},
+		},
+		{
+			name:            "deleted VAC with finalizer but it's referenced by a PV -> finalizer is not removed",
+			initialObjects:  []runtime.Object{pv1WithVAC1},
+			updatedVAC:      vac1TerminatingWithFinalizer,
+			expectedActions: []clienttesting.Action{},
+		},
+		{
+			name:            "deleted VAC with finalizer but it's referenced by a PVC -> finalizer is not removed",
```

Comment:
Nit: should this be "deleting VAC" instead of "deleted VAC"?

Reply:
will do